Storage/SQL/SQLiteHistory.swift

```diff
@@ -941,29 +941,37 @@ extension SQLiteHistory: SyncableHistory {
 
         // We'll need to runQueryUnsafe so we get a LiveCursor, i.e. we don't get the cursor
         // contents into memory all at once.
-        return db.runQueryUnsafe(sql, args: nil, factory: rowIdentity) >>== { cursor in
-            defer { cursor.close() }
-
+        return db.runQueryUnsafe(sql, args: nil, factory: rowIdentity) { (cursor: Cursor<SDRow>) ->  [Int: [Visit]] in
+            // Accumulate a mapping of site IDs to list of visits. Each list should be less than visitLimit.
             // Seed our accumulator with empty lists since we already know which IDs we will be fetching.
             var visits = [Int: [Visit]]()
             historyIDs.forEach { visits[$0] = [] }
 
             // We need to iterate through these explicitly, without relying on the
             // factory.
             for row in cursor.makeIterator() {
```

Comment:
You should check here whether the cursor is in `CursorStatus.failure` and signal failure. That means you'll need to slightly extend the signature of `block` here (perhaps to return `Result<[…]>`).

The reason for this is to avoid a transient or permanent SQLite error from causing dozens of empty history records from being uploaded, as we fail to fetch their visits here.